data_structures/segment_tree.c

```diff
@@ -0,0 +1,170 @@
+
+/**
+ * Compare function
+ * 
+ * @return >0 number if a > b, 0 if a = b, <0 number if a < b
+*/
+typedef int (*comp_func_t)(const void *a, const void *b);
+
+/**
+ * Operation function
+ * 
+ * @param a first element
+ * @param b second element
+ * @param result pointer to memory where result of operation should be written in
+*/
+typedef void (*oper_func_t)(const void *a, const void *b, void *result);
+
+/**
+ * Initialization of segment tree
+ * 
+ * @param tree segment tree element pointer
+ * @param size size of array
+ * @param elem_size size of element
+ * @param data array of elements
+*/
+void segment_tree_init(struct segment_tree *tree, size_t size, size_t elem_size, void *data,
+                         comp_func_t comp_func, oper_func_t oper_func);
+
+/**
+ * Builds tree from provided data in segment_tree_init
+ * 
+ * @param tree segment tree element pointer
+*/
+void segment_tree_build(struct segment_tree *tree);
+
+/**
+ * Updates the element of given position to new_elem
+ * 
+ * @param tree segment tree element pointer
+ * @param pos position where to update with new_elem
+ * @param new_elem new value for position pos
+*/
+void segment_tree_update(struct segment_tree *tree, size_t pos, void *new_elem);
+
+/**
+ * Making query on given segment [l, r]
+ * 
+ * @param tree segment tree element pointer
+ * @param l left end of the segment
+ * @param r right end of the segment
+ * @param result pointer to which query result data should be written in
+*/
+void segment_tree_query(struct segment_tree *tree, size_t l, size_t r, void *result);
+
+/**
+ * Disposes memory allocated for segment tree
+ * 
+ * @param tree tree of which memory should be disposed
+*/
+void segment_tree_dispose(struct segment_tree *tree);
+
+/**
+ * Structure of segment tree
+*/
+struct segment_tree
+{
+    comp_func_t comp_func;      /**< compare function */
+    oper_func_t oper_func;      /**< operation function */
+    size_t elem_size;           /**< size of single element */
+    size_t size;                /**< size of data */
+    void *data;                 /**< data of tree */
+    void *t;                    /**< segment tree */
+};
+
+
+/**
+ * Initialization of segment tree
+ * 
+ * @param tree segment tree element pointer
+ * @param size size of array
+ * @param elem_size size of element
+ * @param data array of elements
+*/
+void segment_tree_init(struct segment_tree *tree, size_t size, size_t elem_size, void *data,
+                         comp_func_t comp_func, oper_func_t oper_func){
+    tree -> comp_func = comp_func;
+    tree -> oper_func = oper_func;
+    tree -> size = size;
+    tree -> elem_size = elem_size;
+    tree -> data = malloc(elem_size * size);
+    tree -> t = malloc(4 * size * elem_size);
+    memcpy(tree -> data, data, elem_size * size);
+}
+
+/**
+ * Builds tree from provided data in segment_tree_init
+ * 
+ * @param tree segment tree element pointer
+*/
+void segment_tree_build(struct segment_tree *tree) {
```

Comment:
where is the code to build the segment tree?
this function seems to be empty

Reply:
As you can see I have written in the pull request review that there is only headers of functions with comments and I will continue to work on it make it complete. I did so because it's not gonna be such an easy task to write generic segment tree also fully test it in C and if someone wants to help me in the task I will accept any helping hand. @ayaankhan98

Reply:
> As you can see I have written in the pull request review that there is only headers of functions with comments and I will continue to work on it make it complete. I did so because it's not gonna be such an easy task to write generic segment tree also fully test it in C and if someone wants to help me in the task I will accept any helping hand. @ayaankhan98

okay keep working, and mark this pull request as a draft. when it is ready for review ping any of the group members for review.
if you need any kind of help regaring this PR feel free to ping me.

Reply:
I think if this code is merged than someone willing to write will fork and write, so I think merging the code would be better

Reply:
Take your time @vakhokoto. There is no rush to complete the add. But Adding a complete and working code would be better to avoid broken repo.

Reply:
Okey, Krishna